### ISSUE TITLE
Allow FundPsbt/PublishTransaction to be used without double-spend risk

### ISSUE THREAD

**Issue by joostjager** [lightningnetwork/lnd#5425]
Currently `FundPsbt` locks the tx inputs for ten minutes. The (finalized) transaction needs to be published within that time frame in order to be sure that no double-spend will happen. But it seems that there is no way to guarantee this?

If the call to `PublishTransaction` is initiated in time, but the lock duration is exceeded before lnd processes the call (for example because of network issues), it could happen that in the mean time another process used those locked utxos, resulting in a double-spend.

**Comment by wpaulino** [comment]
Since the lock duration is configurable now, you could lock inputs for a very long duration and unlock them if the broadcast fails (assuming a full node backend).

**Comment by joostjager** [comment]
I think `FundPsbt` requires that all inputs provided are not locked yet?

**Comment by wpaulino** [comment]
It does, so if we allow a custom duration there as well then it seems this risk is no longer relevant?

**Comment by joostjager** [comment]
Yes a custom duration, but I guess we'd then also need to ability to release the inputs via rpc. Iirc that is currently not possible for 'internally' locked inputs.

**Comment by kaloudis** [comment]
Started some preliminary work on this here: https://github.com/lightningnetwork/lnd/pull/5935. @joostjager do you have any more context as to why internally locked inputs cannot be released via rpc?

**Comment by joostjager** [comment]
@kaloudis internal locks are handled differently, but don't know the background of that.

**Comment by guggero** [comment]
According to the comment [here](https://github.com/lightningnetwork/lnd/blob/master/lnrpc/walletrpc/walletkit_server.go#L389) you can unlock internal leases but not re-lock them before they expire.

I think the original intent was that you couldn't interfere with internal stuff. But PSBTs aren't that internal, so maybe it would make sense to be able to specify a custom lock ID in the `FundPsbt` call?

**Comment by guggero** [comment]
Hmm, you're right, that could lead to timing related issues. I guess a fix would be to be able to specify your own lock ID when using the PSBT API?
